Add documentation for sequences (#8496)
cvc5/cvc5: docs/theories/sequences.rst

Theory Reference: Sequences
===========================

.. note::
  cvc5 currently only supports sequences where the element sort either has an
  infinite domain, e.g., sequences of integers, or a finite domain of a fixed
  cardinality, e.g. bit-vectors.

Semantics
^^^^^^^^^

.. code-block::

  * (seq.empty (Seq S))

    ⟦seq.empty⟧ = []

  * (seq.unit S (Seq S))

    ⟦seq.unit⟧(x) = [x]

  * (seq.len (Seq S) Int)

>     ⟦seq.len⟧(s) is the number of elements in the sequence s, denoted as |s|.

Reviewer: ```suggestion
    ⟦seq.len⟧(s) is the length of the sequence s, denoted as |s|.
```

                    i.e. i < 0 or i >= |s|.

  * (seq.update ((Seq S) Int (Seq S)) (Seq S))

    ⟦seq.update⟧(s, i, sub) is a sequence obtained by updating the continuous
                            sub-sequence of s starting at index i by sub.
                            The updated sequence has the same length as |s|.
                            If i + |sub| > |s|,
                            the out of bound part of sub is ignored.
                            If i out of bound, i.e. i < 0 or i >= |s|,
                            the updated sequence remains same with s.
  
  * (seq.extract ((Seq S) Int Int) (Seq S))

    ⟦seq.extract⟧(s, i, j) is the maximal sub-sequence of s that starts at
                           index i and has length at most j,
                           in case both i and j are non-negative and i is
                           smaller than |s|.
                           Otherwise, the return value is the empty sequence.

   * (seq.++ ((Seq S) (Seq S)) (Seq S))

    ⟦seq.++⟧(s1, s2) is a sequence that is the concatenation of s1 and s2.

   * (seq.at ((Seq S) Int) (Seq S))

    ⟦seq.at⟧(s, i) is a unit sequence that contains the i-th element of s as

>                    the only element.

Reviewer: ```suggestion
                   the only element, or is the empty sequence if i < 0 or i > |s|.
```

   * (seq.contains ((Seq S) (Seq S)) Bool)

    ⟦seq.contains⟧(s, sub) is true if sub is a continuous sub-sequence of s,
                           i.e. sub = seq.extract(s, i, j) for some i, j,
                           and false if otherwise.
  
   * (seq.indexof ((Seq S) (Seq S) Int) Int)

    ⟦seq.indexof⟧(s, sub, i) is the first position of sub at or after i in s,
                             and -1 if there is no occurrence.

   * (seq.replace ((Seq S) (Seq S) (Seq S)) (Seq S))

    ⟦seq.replace⟧(s, src, dst) is the sequence obtained by replacing the
                               first occurrence of src by dst in s.

>                                It equals to s if there is no occurrence.

Reviewer: ```suggestion
                               It is s if there is no occurrence.
```
